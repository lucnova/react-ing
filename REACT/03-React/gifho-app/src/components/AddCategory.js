import React, { useState } from 'react';
import PropTypes from 'prop-types';

const AddCategory = ({ setCategories }) => {
	const [inputValue, setInputValue] = useState(''); // Si se deja vacio, el valor es undefined y handleChange Colapsa

	const handleInputChange = (e) => {
		setInputValue(e.target.value);
	};

	const handleSubmit = (e) => {
		e.preventDefault();

		const trimmedValue = inputValue.trim();

		if (trimmedValue.length > 2) {
			setCategories((currCategories) => {
				if (currCategories.findIndex((item) => item === trimmedValue) === -1) {
					setInputValue(''); // Limpiar Input
					return [trimmedValue, ...currCategories];
				} else {
					return currCategories;
				}
			});
		}
	};

	return (
		<>
			<p id="inputValueCheck" hidden>{inputValue}</p>

			<form onSubmit={handleSubmit}>
				<input type="text" value={inputValue} placeholder="s e a r c h" onChange={handleInputChange}></input>
				<button>o k</button>
			</form>
		</>
	);
};

AddCategory.propTypes = {
	setCategories: PropTypes.func.isRequired,
};

export default AddCategory;
