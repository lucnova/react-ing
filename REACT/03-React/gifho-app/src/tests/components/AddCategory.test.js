import { shallow } from 'enzyme';
import AddCategory from '../../components/AddCategory';

describe('Testing on AddCategory', () => {
	const setCategoriesFn = jest.fn();
	let wrapper = shallow(<AddCategory setCategories={setCategoriesFn} />);

	beforeEach(() => {
		/* Limipiar las simulaciones */
		jest.clearAllMocks();
		wrapper = shallow(<AddCategory setCategories={setCategoriesFn} />);
	});

	test('should display component as snapshot', () => {
		expect(wrapper).toMatchSnapshot();
	});

	test('should display changed value inside input text', () => {
		const sentValue = 'Hey';
		const input = wrapper.find('input[type="text"]');

		input.simulate('change', { target: { value: sentValue } });

		expect(wrapper.find('#inputValueCheck').text().trim()).toBe(sentValue);

		/*
        ESTO DA ERROR:
        
		const inputValue = wrapper.find('#inputValueCheck');
        ...
        expect(inputValue.text().trim()).toBe(sentValue);
        */
	});

	test('should NOT call setCategories WITHOUT a value', () => {
		wrapper.find('form').simulate('submit', { preventDefault() {} });

		expect(setCategoriesFn).not.toHaveBeenCalled();
	});
});
