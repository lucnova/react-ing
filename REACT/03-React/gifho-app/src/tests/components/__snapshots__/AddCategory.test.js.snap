// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing on AddCategory should display component as snapshot 1`] = `
<form
  onSubmit={[Function]}
>
  <input
    onChange={[Function]}
    placeholder="s e a r c h"
    type="text"
    value=""
  />
  <button>
    o k
  </button>
</form>
`;
